### PR TITLE
Work

### PR DESCRIPTION
**Pull request by alexisbarraza3293**

### CODE REVIEW

File: util/dist.h [hunk 1]
```diff
@@ -20,43 +20,52 @@
...
+	private:
+	/*We do not have span so we are using a raw pointer and size_t to describe
+	our distribution.*/
+	const dist_point* dist_point_ptr = nullptr;
+	int dist_size = 0;
+	size_t max_message_length = 0;
+	double dist_mean = 0;
+	double comp_dist_mean();
+	int fixed_size = 0;
+	std::unique_ptr<dist_point> fixed_dist;
```
Comment (danmanj) [inline util/dist.h:69]: make this a member variable

File: util/dist.cc [hunk 2]
```diff
@@ -221,33 +193,55 @@ int dist_msg_overhead(int length, int mtu)
...
+	std::vector<int> output;
+	output.reserve(dist_size + 1); //accomodate for max_msg_length
+	bool max_msg_len_used = false;
+
+	for (int i = 0; i < dist_size; ++i) {
+		if(dist_point_ptr[i].length < max_message_length) {
+			output.push_back(dist_point_ptr[i].length);
+		}
+		else {
+			max_msg_len_used = true;
```
Comment (danmanj) [inline util/dist.cc:230]: push max and break

File: util/dist.cc [hunk 3]
```diff
@@ -221,33 +193,55 @@ int dist_msg_overhead(int length, int mtu)
...
 	overhead = 0.0;
 	prev_fraction = 0.0;
-	for (dist_point &p: points) {
-		overhead += static_cast<double>(dist_msg_overhead(p.length, mtu))
-				* (p.fraction - prev_fraction)
-				/ static_cast<double>(p.length);
-		prev_fraction = p.fraction;
-		if (p.fraction >= 1.0)
+	for (int i = 0; i < dist_size; ++i) {
+		overhead += static_cast<double>(dist_msg_overhead(dist_point_ptr[i].length, mtu))
```
Comment (danmanj) [inline util/dist.cc:205]: static cast is not necessary

File: util/dist.cc [hunk 4]
```diff
@@ -60,61 +57,34 @@ dist_point *dist_lookup(const char *dist, int *fixed)
...
-	for (int i = 0; i < num_samples; i++) {
-		double cdf_fraction = uniform_dist(*rand_gen);
-		for (dist_point &p: points) {
-			if (p.fraction > cdf_fraction) {
-				sizes.push_back(p.length);
-				break;
-			}
-		}
-	}
+	//should not reach here
```
Comment (danmanj) [inline util/dist.cc:60]: retyurn null, move error handling back to caller

File: util/dist.cc [hunk 5]
```diff
@@ -18,10 +18,18 @@
...
 #include "dist.h"
 
+//Forward declairing these arrays of CDFs (defined at the bottom).
+extern dist_point_gen::dist_point w1[];
+extern dist_point_gen::dist_point w2[];
+extern dist_point_gen::dist_point w3[];
+extern dist_point_gen::dist_point w4[];
+extern dist_point_gen::dist_point w5[];
+
 /**
```
Comment (danmanj) [inline util/dist.cc:33]: remove docs about fixed

File: util/Makefile [hunk 6]
```diff
@@ -1,10 +1,10 @@
 # Makefile to build various testing programs for Homa.
 
-CFLAGS := -Wall -Werror -fno-strict-aliasing -O3 -I..
+CFLAGS := -Wall -Werror -fno-strict-aliasing -O0 -I..
```
Comment (danmanj) [inline util/Makefile:3]: -O3

File: util/dist.h [hunk 7]
```diff
@@ -20,43 +20,55 @@
...
+	};
+
+	private:
+	/*We do not have span so we are using a raw pointer and size_t to describe
+	our distribution.*/
+	const dist_point* dist_point_ptr = nullptr;
+	int dist_size = 0;
+	//The maximum size of a message length passed through the constructor
+	size_t max_message_length = 0;
+	//Computes the mean of the distribution and stores it in dist_mean
```
Comment (danmanj) [inline util/dist.h:67]: this doesn't store the mean, it just computes it.

File: util/dist.cc [hunk 8]
```diff
@@ -129,56 +94,58 @@ double dist_mean(std::vector<dist_point> &points)
...
  *               variation in size.
- *
- * Return:       A (possibly condensed) distribution, with points in increasing
- *               order of @length. If the result is empty, it means that @dist
- *               wasn't a valid name.
  */
-std::vector<dist_point> dist_get(const char *dist, int max_length,
-		double min_bucket_frac, double max_size_range)
+dist_point_gen::dist_point_gen(const char* dist, size_t max_length, double min_bucket_frac,
+			  double max_size_range)
```
Comment (danmanj) [inline util/dist.cc:99]: this should line up with the line above

File: util/dist.cc [hunk 9]
```diff
@@ -129,56 +94,58 @@ double dist_mean(std::vector<dist_point> &points)
...
-			last_buck_fraction = 1.0;
-			break;
-		}
-		if (buck_short < 0)
-			buck_short = p->length;
-		if (p->length > max_size_range*buck_short) {
-			result.emplace_back(prev->length, prev->fraction);
-			last_buck_fraction = prev->fraction;
-			buck_short = p->length;
+	else {
```
Comment (danmanj) [inline util/dist.cc:112]: else should be on the previous line

File: util/dist.cc [hunk 10]
```diff
@@ -129,56 +94,58 @@ double dist_mean(std::vector<dist_point> &points)
...
+		dist_point_ptr = dist_lookup(dist);
+		size_t i = 0;
```
Comment (danmanj) [inline util/dist.cc:114]: rewrite without an extra temporary "i" variable.

File: util/dist.cc [hunk 11]
```diff
@@ -221,33 +188,52 @@ int dist_msg_overhead(int length, int mtu)
...
-				* (p.fraction - prev_fraction)
-				/ static_cast<double>(p.length);
-		prev_fraction = p.fraction;
-		if (p.fraction >= 1.0)
+	for (int i = 0; i < dist_size; ++i) {
+		overhead += dist_msg_overhead(dist_point_ptr[i].length, mtu)
+				* (dist_point_ptr[i].fraction - prev_fraction)
+				/ dist_point_ptr[i].length;
+		prev_fraction = dist_point_ptr[i].fraction;
+		if (dist_point_ptr[i].fraction >= 1.0)
```
Comment (danmanj) [inline util/dist.cc:204]: this should not be possible.

File: util/dist.cc [hunk 12]
```diff
@@ -221,33 +188,52 @@ int dist_msg_overhead(int length, int mtu)
...
+ */
+std::vector<int> dist_point_gen::sizes() const
+{
+	std::vector<int> output;
+	output.reserve(dist_size + 1); //accomodate for max_msg_length
+
+	for (int i = 0; i < dist_size; ++i) {
+		if(dist_point_ptr[i].length < max_message_length) {
+			output.push_back(dist_point_ptr[i].length);
+		}
```
Comment (danmanj) [inline util/dist.cc:222]: else should be on previous line

**Comment by johnousterhout** [comment]
Is this PR needed anymore?

**Pull request closed without merging**